src/storage.py

```diff
@@ -0,0 +1,267 @@
+
+    def _enroll_disks_in_batch(self, disks: list):
+        """Adds requested Disks to Microceph and stored state."""
+        microceph.enroll_disks_as_osds(disks)
+        for disk in disks:
+            self._save_osd_data(disk)
+        logger.debug(f"Added {disks} as OSDs.")
+
+    def remove_osd(self, osd: int, force: bool = False):
+        """Removes OSD from MicroCeph and from stored state."""
+        try:
+            microceph.remove_disk_cmd(osd, force)
+            self._remove_osd_data(self._stored.osd_data[osd])
+        except CalledProcessError as e:
+            if force:
+                # clean stored state.
+                self._remove_osd_data(self._stored.osd_data[osd])
+            raise e
+
+    def _save_osd_data(self, disk: str, wal: str = None, db: str = None):
+        """Save OSD data."""
+        for osd in microceph.list_disk_cmd()["ConfiguredDisks"]:
+            # e.g. check 'vdd' in '/dev/vdd'
+            if str(microceph._get_disk_info(osd["path"], "name")) in disk:
+                self._stored.osd_data[osd["osd"]] = {
+                    "disk_by_id": osd["path"],
+                    "disk": disk,
+                    "wal": wal,
+                    "db": db,
+                }
+
+    def _get_osd_num(self, disk, directive):
+        """Fetch the OSD number of consuming OSD, None is not used as OSD."""
+        # both osd-devices and disks are used as OSD disks.
+        if directive in ["osd-devices", "disk"]:
+            directive = "disk"
+
+        logger.debug(self._stored.osd_data)
+        logger.debug(f"Incoming Disk {disk}, directive {directive}.")
+
+        for k, v in dict(self._stored.osd_data).items():
+            if v and v[directive] == disk:
+                return k  # key is the stored osd number.
+        return None
+
+    def _clean_stale_osd_data(self):
+        """Compare with disk list and remove stale entries."""
+        osds = [osd["osd"] for osd in microceph.list_disk_cmd()["ConfiguredDisks"]]
+
+        for key in dict(self._stored.osd_data).keys():
+            if key not in osds:
+                val = self._stored.osd_data[key]
+                self._stored.osd_data[key] = None
+                logger.debug(f"Popped {val}")
+
+    def _remove_osd_data(self, disk: str):
+        """Remove data for removed OSD."""
+        num = -1  # impossible osd number.
+        for osd_num, data in dict(self._stored.osd_data).items():
+            if data and data["disk"] == disk:
+                num = osd_num
+        if num > 0:
+            val = self._stored.osd_data[num]
+            self._stored.osd_data[num] = None
+            logger.debug(f"Popped {val}")
+
+        logger.debug(self._stored.osd_data)
+
+    # NOTE(utkarshbhatthere): 'storage-get' sometimes fires before
+    # requested information is available.
+    @retry(wait=wait_fixed(5), stop=stop_after_attempt(10))
+    def juju_storage_get(self, storage_id=None, attribute=None):
```

Comment:
Doesn't the ops framework have bindings for this (and `juju_storage_set` as well)?

Reply:
There indeed exists an implementation for get and list in `_ModelBacked Class' [here](https://ops.readthedocs.io/en/latest/_modules/ops/model.html#). But they are not exposed and the doc line says: `Charm authors should not directly interact with the ModelBackend, it is a private
    implementation of Model.`

Reply:
To add over this, those implementation run the same commands in hook/event context and needed some sort of wrapping to be easily usable (retries, exception handling etc) so I decided to write these out instead.